learn/functions/global_variables.md

```diff
@@ -0,0 +1,177 @@
+        
+    inner()
+    print("outer:", x)
+
+outer()  # output: inner: nonlocal, outer: nonlocal
+```
+
+In this example, `x` is a local variable in the `outer` function. The `inner` function has access to this variable using the `nonlocal` keyword, so we can modify its value. When we call the `outer` function, the `inner` function is executed and the value of `x` is changed to `nonlocal`. This change is reflected in the `outer` function when we `print` the value of `x` after the `inner` function is executed.  
+  
+## Best practices for using global variables in function  
+
+Global variables are variables that can be accessed and modified from any part of the program. In Python, a variable's scope determines where it can be accessed. The best practice for using global variables in Python is to minimize their usage, as too many global variables can make the program difficult to understand, debug, and maintain.
+
+One example of a global variable is the power function in Python. We can use a loop to calculate the power of a number. Here is an example:
+
+```python
+power = 1
+
+def calculate_power(num, exponent):
+    global power
+    power = 1
+    for i in range(exponent):
+        power *= num
+    return power
+
+print("Power of 2^3 is", calculate_power(2, 3))
+print("Power of 5^4 is", calculate_power(5, 4))
+```
+
+In this example, we declare a global variable `power` outside the function `calculate_power()`. Inside the function, we use this global variable to store the power of the number. We reset the value of the `power` variable to 1 for each new calculation.
+
+### Program to count the number of times a function is called:
+
+```python
+count = 0
+
+def my_function():
+    global count
+    count += 1
+    print("This function has been called", count, "times.")
+
+my_function()
+my_function()
+my_function()
+```
+
+In this example, we declare the global variable `count` outside the function `my_function()`. Inside the function, we increment the value of the `count` variable every time the function is called. We then print the value of `count`. 
+
+Overall, it is generally best to avoid global variables in favor of local variables with more limited scope. However, in some cases, global variables may be necessary or useful, and we can use them carefully with the `global` keyword to access them inside functions.  
+  
+## Advanced tips for python variable scopes and naming conventions  
+
+Advanced tips for Python variable scopes include avoiding global variables as much as possible to prevent naming conflicts and unexpected behavior. It is also recommended to use descriptive variable names that follow the PEP 8 naming conventions, such as using lowercase letters and underscores to separate words.
+
+Function naming conventions in Python follow the same PEP 8 guidelines, using lowercase letters and underscores to separate words. Function names should also be descriptive and convey the purpose of the function.
```

Comment:
If we reference to some external documentation - it's good to insert an HTML link to the source. But add links only to the trusted official sources.

Reply:
In this case I mean "PEP 8 guidelines".